internal/service/resources/collector/utils.go

```diff
@@ -0,0 +1,225 @@
+	switch {
+	case errors.Is(err, utils.ErrNotFound):
+		// New object instance
+		after, err = utils.Create[T](ctx, db, object)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create object '%s/%s': %w", object.TableName(), uuid, err)
+		}
+
+		slog.Debug("object inserted", "table", object.TableName(), "uuid", uuid, "record", after)
+
+	case err == nil:
+		// Updated object instance
+		before = record
+
+		// We only need to update the fields that have actually changed so compare them to get the list of fields.
+		tags := utils.CompareObjects(*record, object, "CreatedAt")
+		if len(tags) == 0 {
+			// This shouldn't happen since the generation id always changes
+			after = before
+			slog.Warn("no change detected on persisted object", "table", object.TableName(), "uuid", uuid)
+			return before, after, nil
+		}
+
+		after, err = utils.Update[T](ctx, db, uuid, object, tags.Fields()...)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to update object '%s/%s': %w", object.TableName(), uuid, err)
+		}
+
+		slog.Debug("object updated",
+			"table", object.TableName(), "uuid", uuid, "before", before, "after", after, "columns", tags.Fields())
+	default:
+		return nil, nil, fmt.Errorf("failed to get object '%s/%s': %w", object.TableName(), uuid, err)
+	}
+
+	return before, after, nil
+}
+
+// persistDataChangeEvent persists a data change object to its database table.  The before and
+// after model objects are marshaled to JSON prior to being stored.
+func persistDataChangeEvent(ctx context.Context, db *pgxpool.Pool, tableName string, uuid uuid.UUID,
+	parentUUID *uuid.UUID, before, after any) (*models.DataChangeEvent, error) {
+	var err error
+	var beforeJSON, afterJSON []byte
+	if before != nil {
+		if beforeJSON, err = json.Marshal(before); err != nil {
+			return nil, fmt.Errorf("failed to marshal before object: %w", err)
+		}
+	}
+	if after != nil {
+		if afterJSON, err = json.Marshal(after); err != nil {
+			return nil, fmt.Errorf("failed to marshal after object: %w", err)
+		}
+	}
+
+	dataChangeEvent := models.DataChangeEvent{
+		ObjectType: tableName,
+		ObjectID:   uuid,
+		ParentID:   parentUUID,
+	}
+
+	if beforeJSON != nil {
+		value := string(beforeJSON)
+		dataChangeEvent.BeforeState = &value
+	}
+	if afterJSON != nil {
+		value := string(afterJSON)
+		dataChangeEvent.AfterState = &value
+	}
+
+	result, err := utils.Create[models.DataChangeEvent](ctx, db, dataChangeEvent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create data change event: %w", err)
+	}
+
+	return result, nil
+}
+
+// persistObjectWithChangeEvent persists an object to its database table and if the external API
+// model representation of the object has changed then a data change event is stored.  Persisting
+// of the object and its change event are captured under the same transaction to ensure we never
+// lose any change events.
+func persistObjectWithChangeEvent[T db.Model](ctx context.Context, db *pgxpool.Pool, record T,
```

Comment:
What do you think about creating methods in dbModel to return the primarykey and parentId values? It could simplify the signature of these methods. T.GetprimaryKey()

Reply:
Maybe in a future PR.  I was hoping to avoid adding too much to the dbModel.  For now this two extra parameters are used in many places.